File: leetcode106.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        return self.seperate(inorder, postorder)

    def seperate(self, inorder, postorder):
        if len(postorder) == 0:
            return None
        root = TreeNode(postorder[-1])
        idx = inorder.index(postorder[-1])
        lp,li = len(postorder), len(inorder)
        root.left = self.seperate(postorder[0:idx], inorder[0:idx])
        root.right = self.seperate(postorder[idx:lp-1], inorder[idx+1:li])
        return root

File: leetcodec++/leetcode106.cpp
//
// Created by 黄敏辉 on 2019-08-01.
//

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return helper(postorder, 0, postorder.size(), inorder, 0, inorder.size());
    }

    TreeNode* helper(vector<int>& postorder, int i, int j, vector<int>& inorder, int k, int l){
        if (i >= j ||k >= l)
            return NULL;

        auto f = find(inorder.begin() + k, inorder.begin() + l, postorder[j-1]);

        int dis = f - inorder.begin() - k;

        TreeNode* root = new TreeNode(postorder[j-1]);
        root->left = helper(postorder, i, i+dis, inorder, k, k+dis);
        root->right = helper(postorder, i+dis,j-1, inorder ,k+dis+1, l);
        return root;
    }
};